homeassistant/components/template/weather.py

```diff
@@ -327,4 +367,66 @@ async def async_added_to_hass(self) -> None:
                 "_forecast",
                 self._forecast_template,
             )
+
+        if self._forecast_daily_template:
+            self.add_template_attribute(
+                "_forecast_daily",
+                self._forecast_daily_template,
+                on_update=self._update_daily,
+                validator=self._validate_forecast,
+            )
+        if self._forecast_hourly_template:
+            self.add_template_attribute(
+                "_forecast_hourly",
+                self._forecast_hourly_template,
+                on_update=self._update_hourly,
+                validator=self._validate_forecast,
+            )
+        if self._forecast_twice_daily_template:
+            self.add_template_attribute(
+                "_forecast_twice_daily",
+                self._forecast_twice_daily_template,
+                on_update=self._update_twice_daily,
+                validator=self._validate_forecast,
+            )
+
         await super().async_added_to_hass()
+
+    @callback
+    def _update_daily(self, result: list[Forecast] | TemplateError) -> None:
+        """Save template result and trigger forecast listener."""
+        attr_result = None if isinstance(result, TemplateError) else result
+        setattr(self, "_forecast_daily", attr_result)
+        self.hass.create_task(self.async_update_listeners(["daily"]))
+
+    @callback
+    def _update_hourly(self, result: list[Forecast] | TemplateError) -> None:
+        """Save template result and trigger forecast listener."""
+        attr_result = None if isinstance(result, TemplateError) else result
+        setattr(self, "_forecast_hourly", attr_result)
+        self.hass.create_task(self.async_update_listeners(["hourly"]))
+
+    @callback
+    def _update_twice_daily(self, result: list[Forecast] | TemplateError) -> None:
+        """Save template result and trigger forecast listener."""
+        attr_result = None if isinstance(result, TemplateError) else result
+        setattr(self, "_forecast_twice_daily", attr_result)
+        self.hass.create_task(self.async_update_listeners(["twice_daily"]))
+
+    @callback
+    def _validate_forecast(
+        self, result: list[Forecast] | TemplateError
+    ) -> list[Forecast] | None:
+        """Validate the forecasts."""
+        if result is None or isinstance(result, TemplateError):
+            return None
+        check_keys = Forecast.__annotations__.keys()
+        if isinstance(result, list):
+            for forecast in result:
+                for key, _ in forecast.items():
+                    if key not in check_keys:
```

Comment:
I think we can use set operations instead, we should check that the keys of the current element is a subset of all allowed keys.
This should also check that each element has `datetime` as well as `is_daytime` for `twice_daily`forecasts